packages/cosmic-swingset/src/chain-main.js

```diff
@@ -401,6 +408,65 @@ export default async function main(progname, args, { env, homedir, agcc }) {
     const keepSnapshots =
       XSNAP_KEEP_SNAPSHOTS === '1' || XSNAP_KEEP_SNAPSHOTS === 'true';
 
+    const nodeHeapSnapshots = Number.parseInt(NODE_HEAP_SNAPSHOTS, 10);
+
+    let lastSnapshotTime = nodeHeapSnapshots >= 0 ? 0 : NaN;
+    let commitCallsSinceLastSnapshot = nodeHeapSnapshots > 0 ? -1 : NaN;
+    const snapshotBaseDir = path.resolve(stateDBDir, 'node-heap-snapshots');
+
+    let heapSnapshot;
+
+    if (nodeHeapSnapshots >= 0) {
+      fs.mkdirSync(snapshotBaseDir, { recursive: true });
+    }
+
+    const getRuntimeStats = () => {
+      const t0 = performance.now();
+      const heapStats = v8.getHeapStatistics();
+      const t1 = performance.now();
+      const heapSpaceStats = v8.getHeapSpaceStatistics();
+      const t2 = performance.now();
+      const memoryUsage = process.memoryUsage();
+      const t3 = performance.now();
+
+      commitCallsSinceLastSnapshot += 1;
+      if (
+        t0 - lastSnapshotTime > 30 * 1000 ||
+        (nodeHeapSnapshots > 0 &&
+          commitCallsSinceLastSnapshot >= nodeHeapSnapshots)
+      ) {
+        commitCallsSinceLastSnapshot -= commitCallsSinceLastSnapshot;
+        lastSnapshotTime = t0;
+        heapSnapshot = `Heap-${process.pid}-${Date.now()}.heapsnapshot`;
+      } else {
+        heapSnapshot = undefined;
+      }
+
+      return {
+        heapStats,
+        heapSpaceStats,
+        memoryUsage,
+        heapSnapshot,
+        statsTime: {
+          heapStats: t1 - t0,
+          heapSpaceStats: t2 - t1,
+          memoryUsage: t3 - t2,
+        },
+      };
+    };
+
+    const afterCommitCallback = async () => {
+      await waitUntilQuiescent();
+
+      engineGC();
```

Comment:
This is technically a behavior change forcing a full GC after committing every block.

Reply:
But it's only affecting the behavior of the kernel, not the workers, and the kernel doesn't use WeakMaps or WeakRefs or anything that would allow GC to affect visible behavior (it should only affect timing).

The `waitUntilQuiescent()` is kinda weird in a kernel context, though. It makes sense in a worker because we deny `setImmediate()`/etc to userspace, so there's a distinction between the kind of code that loses agency before this fires, and the kind of code that *could* retain agency (but doesn't, because it's all written by us). I'm not sure what it means in the kernel world.

Nothing in the kernel/controller uses `setImmediate()`, but all the vat managers use IO, and `waitUntilQuiescent` will fire before IO returns. I think I'd want to make sure this isn't run until after `controller.run()` has returned (at which point all the workers should be idle, and no V8 IO should be happening). Given that, it might manage to delay the GC until after some lingering promises are running. But my hunch is that it'd be better/simpler/no-worse to just call `engineGC()` without the `waitUntilQuiescent()`.

Reply:
> The `waitUntilQuiescent()` is kinda weird in a kernel context, though

It's to make sure that all reactions to inform the go side have been processed before we start this work. But I agree the wording is not right (only the implementation using `setImmediate`)

> But my hunch is that it'd be better/simpler/no-worse to just call `engineGC()` without the `waitUntilQuiescent()`.

Nope, there are some promise reactions that need to happen before we do `replier.resolve()` in `fromGo`. This is basically what the `setImmediate` is for here, make sure all these promise reactions (which have no dependency on I/O) are effectively over. I actually considered using `process.nextTick()` since that uses a parallel queue to the promise queue of equal priority, but that isn't really a guarantee in future versions of node.

Reply:
> But it's only affecting the behavior of the kernel, not the workers, and the kernel doesn't use WeakMaps or WeakRefs or anything that would allow GC to affect visible behavior (it should only affect timing).

Correct this is only a timing change for the kernel.